src/ccontrol/UserQuerySelect.h

```diff
@@ -152,7 +152,10 @@ class UserQuerySelect : public UserQuery {
 
     void _expandSelectStarInMergeStatment(std::shared_ptr<query::SelectStmt> const& mergeStmt);
 
-    void _discardMerger();
+    /// Discard the merger if it exists.
+    /// @param lock a lock on the mutex _killMutex that prevents the race condition if the query
+    /// cancellation happens at the same time the query resources are being discarded.
+    void _discardMerger(std::lock_guard<std::mutex> const& lock);
```

Comment:
Calling `.reset()` for various pointers in `discard()` for the children of `UserQuery` is dangerous and probably unnecessary in most cases, as `discard()` is called right before the `uq` is reset in the `finalizer`  in `Czar::submitQuery`, so they are about to be deleted anyway. It makes the code more complicated and the benefit isn't clear.

Reply:
The `.reset()` is already in the code. I don't want to change the logic flow of the existing code. My changes were just to prevent existing code from resetting the pointers while there is a chance that the query cancellation event may arrive. Hence I was very careful (not) to make any further mods to that code. Let's delay any dramatic changes to some significant redesign of this (and similar) classes.